pool/src/lib.rs

```diff
@@ -735,34 +844,134 @@ impl LendingPool {
         mb_account_data: Option<AccountData>,
-        is_good: bool,
     ) -> Result<bool, Error> {
-        let _account_data = if let Some(account_data) = mb_account_data {
+        let account_data = if let Some(account_data) = mb_account_data {
             account_data
         } else {
             let user_config = read_user_config(env, who.clone())?;
             let reserves = read_reserves(env);
-            Self::calc_account_data(env, who, mb_who_balance, &user_config, &reserves)?
+            Self::calc_account_data(env, who, mb_who_balance, &user_config, &reserves, false)?
         };
 
-        Ok(is_good)
+        Ok(account_data.npv > 0)
     }
 
     fn require_good_position(
         env: &Env,
         who: Address,
         mb_who_balance: Option<(Address, i128)>,
         mb_account_data: Option<AccountData>,
-        is_good: bool,
     ) -> Result<(), Error> {
-        let is_good_position =
-            Self::is_good_position(env, who, mb_who_balance, mb_account_data, is_good)?;
+        let is_good_position = Self::is_good_position(env, who, mb_who_balance, mb_account_data)?;
         if !is_good_position {
             return Err(Error::BadPosition);
         }
 
         Ok(())
     }
+
+    fn do_liquidate(
+        env: &Env,
+        liquidator: Address,
+        who: Address,
+        reserves: Vec<Address>,
+        user_config: &mut UserConfiguration,
+        liquidation_debt: i128,
+        get_stoken: bool,
+    ) -> Result<(), Error> {
+        let mut sorted_collat_reserves = Vec::new(env);
+        let mut debt_reserves = Vec::new(env);
+        // soroban_sdk::Vec doesn't have built-in sort or something like binary_search_by_key
+        // so here a little hack that uses second Vec with discounts that tracks indexes for `collat_reserves`
+        let mut sorted_discounts = Vec::new(env);
+        for i in 0..reserves.len() {
+            let asset = reserves.get_unchecked(i).unwrap_optimized();
+            if user_config.is_using_as_collateral_or_borrowing(env, i as u8) {
+                let reserve = read_reserve(env, asset.clone())?;
+                if !reserve.configuration.is_active {
+                    return Err(Error::NoActiveReserve);
+                }
+                if user_config.is_borrowing(env, i as u8) {
+                    debt_reserves.push_back(reserve);
+                } else if user_config.is_using_as_collateral(env, i as u8) {
+                    let idx = match sorted_discounts.binary_search(reserve.configuration.discount) {
+                        Ok(exists) => exists,
+                        Err(new) => new,
+                    };
+                    sorted_discounts.insert(idx, reserve.configuration.discount);
+                    sorted_collat_reserves.insert(idx, reserve);
+                }
+            }
+        }
+
+        let mut debt_to_cover = liquidation_debt;
+        for reserve in sorted_collat_reserves {
+            if debt_to_cover == 0 {
```

Comment:
Is it possible if `debt_to_cover` is 0, but we still have `sorted_collat_reserves` to enumerate?

Reply:
If debt_to_cover=0 then position is good and all liquidatuion will fail with error